lib/libzpool/uzfs_io.c

```diff
@@ -60,6 +64,31 @@ uzfs_write_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 
 	rl = zfs_range_lock(&zv->zv_range_lock, r_offset, r_len, RL_WRITER);
 
+	/*
+	 * TODO: error handling in case of UZFS_IO_TX_ASSIGN_FAIL
+	 */
+	if (zv->in_rebuilding_mode) {
+		if (!is_rebuild)
+			uzfs_add_to_incoming_io_tree(zv, offset, len);
+		if (is_rebuild) {
+			count = uzfs_search_incoming_io_tree(zv, offset,
+			    len, (void **)&chunk_io);
+chunk_io:
```

Comment:
This comment is just to improve readability. Its upto you...take it or ignore it.
<Code snippet>
        if (zv->in_rebuilding_mode) {
                if (!is_rebuild) {
                        uzfs_add_to_incoming_io_tree(zv, offset, len);
                } else {
                        count = uzfs_search_incoming_io_tree(zv, offset,
                            len, (void **)&chunk_io);
                }
        }

chunk_io:
        if (count) {
                node = list_remove_head(chunk_io);
                offset = node->offset;
                len = node->len;
                end = offset + len;
                umem_free(node, sizeof (*node));
                wrote = offset - orig_offset;
                len_in_first_aligned_block = 0;
                count--;
                zv->rebuild_bytes += len;
        }
</Code snippet>

Reply:
uzfs_search_incoming_io_tree may return 0 if incoming_io_tree is NULL. This value is checked further  so i kept this code snippet under if (zv->in_rebuilding_mode) if (is_rebuild) only.
It also makes uzfs_write_data less messy by keeping rebuilding code in few blocks only.